liquidctl/keyval.py

```diff
@@ -87,30 +97,112 @@ def store(self, key, value):
         data = repr(value)
         assert literal_eval(data) == value, 'encode/decode roundtrip fails'
         path = os.path.join(self._write_dir, key)
-        fd, tmp = tempfile.mkstemp(dir=self._write_dir, text=True)
-        with open(fd, mode='w') as f:
+
+        with self._open_with_lock(path, os.O_WRONLY | os.O_CREAT | os.O_TRUNC) as f:
             f.write(data)
-            f.flush()
-        os.replace(tmp, path)
+            f.flush()  # ensure flushing before automatic unlocking
+
         _LOGGER.debug('stored %s=%r (in %s)', key, value, path)
 
+    def load_store(self, key, func):
+        value = None
+        new_value = None
+
+        path = os.path.join(self._write_dir, key)
+
+        # lock the destination as soon as possible
+        with self._open_with_lock(path, os.O_RDWR | os.O_CREAT) as f:
+
+            # still traverse all possible locations to find the current value
+            for base in self._read_dirs:
+                read_path = os.path.join(base, key)
+                if not os.path.isfile(read_path):
+                    continue
+                try:
+                    if os.path.samefile(read_path, path):
+                        # we already have an exclusive lock to this file
+                        data = f.read().strip()
+                        f.seek(0)
+                    else:
+                        with self._open_with_lock(read_path, os.O_RDWR, shared=True) as aux:
+                            data = aux.read().strip()
+
+                    if not data:
+                        continue
+
+                    value = literal_eval(data)
+                    _LOGGER.debug('loaded %s=%r (from %s)', key, value, read_path)
+                except OSError as err:
+                    _LOGGER.warning('%s exists but could not be read: %s', read_path, err)
+                except ValueError as err:
+                    _LOGGER.warning('%s exists but was corrupted: %s', key, err)
+                else:
+                    break
+            else:
+                _LOGGER.debug('no data (file) found for %s', key)
+
+            new_value = func(value)
+
+            data = repr(new_value)
+            assert literal_eval(data) == new_value, 'encode/decode roundtrip fails'
```

Comment:
Well, if the driver tries to store some arbitrary class instance, there's no prior assurance that its `__repr__` is complete (=contains all necessary information for reconstruction) or that it can be parsed as a Python literal.  So I though it was better to check this when saving the data, instead of later, when reading it.

Another factor was that our tests used to work with a simpler __dict__ implementation, so they wouldn't catch stuff like this because there was no actual serialization.

Reply:
Ah that makes sense.